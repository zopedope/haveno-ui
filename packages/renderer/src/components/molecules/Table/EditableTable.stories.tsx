// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useState, useEffect } from "react";
import type { ColumnDef } from "@tanstack/react-table";
import { createTable } from "@tanstack/react-table";
import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { Table } from "./Table";

export default {
  title: "atoms/Table/EditableTable",
  component: Table,
} as ComponentMeta<typeof Table>;

const Template: ComponentStory<typeof Table> = () => {
  return (
    <Table
      table={table}
      data={data}
      columns={columns}
      defaultColumn={defaultColumn}
      onEditableDataChange={(values) => {
        console.log(values);
      }}
    />
  );
};

export const Default = Template.bind({});

Default.args = {};

interface Person {
  firstName: string;
  lastName: string;
  age: number;
  visits: number;
  status: string;
  progress: number;
}

const table = createTable().setRowType<Person>().setTableMetaType<{
  updateData: (rowIndex: number, columnId: string, value: unknown) => void;
}>();
type TableGenerics = typeof table.generics;

const columns = [
  table.createGroup({
    header: "Name",
    footer: (props) => props.column.id,
    columns: [
      table.createDataColumn("firstName", {
        cell: (info) => info.getValue(),
        footer: (props) => props.column.id,
      }),
      table.createDataColumn((row) => row.lastName, {
        id: "lastName",
        cell: (info) => info.getValue(),
        header: () => <span>Last Name</span>,
        footer: (props) => props.column.id,
      }),
    ],
  }),
  table.createGroup({
    header: "Info",
    footer: (props) => props.column.id,
    columns: [
      table.createDataColumn("age", {
        header: () => "Age",
        footer: (props) => props.column.id,
      }),
      table.createGroup({
        header: "More Info",
        columns: [
          table.createDataColumn("visits", {
            header: () => <span>Visits</span>,
            footer: (props) => props.column.id,
          }),
          table.createDataColumn("status", {
            header: "Status",
            footer: (props) => props.column.id,
          }),
          table.createDataColumn("progress", {
            header: "Profile Progress",
            footer: (props) => props.column.id,
          }),
        ],
      }),
    ],
  }),
];

const data: Array<Person> = [
  {
    firstName: "tanner",
    lastName: "linsley",
    age: 24,
    visits: 100,
    status: "In Relationship",
    progress: 50,
  },
  {
    firstName: "tandy",
    lastName: "miller",
    age: 40,
    visits: 40,
    status: "Single",
    progress: 80,
  },
  {
    firstName: "joe",
    lastName: "dirte",
    age: 45,
    visits: 20,
    status: "Complicated",
    progress: 10,
  },
];

const defaultColumn: Partial<ColumnDef<TableGenerics>> = {
  cell: ({ getValue, row: { index }, column: { id }, instance }) => {
    const initialValue = getValue();
    const [value, setValue] = useState(initialValue);

    const onBlur = () => {
      instance.options.meta?.updateData(index, id, value);
    };

    useEffect(() => {
      setValue(initialValue);
    }, [initialValue]);

    return (
      <input
        value={value as string}
        onChange={(e) => setValue(e.target.value)}
        onBlur={onBlur}
      />
    );
  },
};
