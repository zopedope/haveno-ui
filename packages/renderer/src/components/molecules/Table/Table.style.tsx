// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { createStyles } from "@mantine/core";

export const useStyles = createStyles((theme) => ({
  primary: {
    "thead tr": {
      backgroundColor: theme.colors.gray[0],

      th: {
        borderBottomColor: theme.colors.gray[2],
        color: theme.colors.gray[5],
        fontSize: 10,
        fontWeight: 700,
        letterSpacing: "0.05em",
        textTransform: "uppercase",
      },
    },
    "tbody tr": {
      td: {
        borderBottomColor: theme.colors.gray[2],
      },
    },
  },
}));
