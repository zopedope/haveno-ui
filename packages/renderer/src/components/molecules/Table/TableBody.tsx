// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { Fragment } from "react";
import { useTableContext } from "./use-table-context";

export function TableBody() {
  const {
    table,
    props: { rowSubComponent, onRowClick },
  } = useTableContext();

  return (
    <tbody>
      {table.getRowModel().rows.map((row) => (
        <Fragment key={row.id}>
          <tr
            key={row.id}
            onClick={() => {
              row.toggleExpanded();
              onRowClick && onRowClick(row);
            }}
          >
            {row.getVisibleCells().map((cell) => (
              <td
                key={cell.id}
                style={{
                  width: cell.column.getSize(),
                  textAlign: cell.column.columnDef?.meta?.textAlign,
                }}
              >
                {cell.renderCell()}
              </td>
            ))}
          </tr>
          {row.getIsExpanded() && rowSubComponent ? (
            <tr key={`${row.id}-subcomponent`}>
              <td colSpan={row.getVisibleCells()?.length}>
                {rowSubComponent({ row })}
              </td>
            </tr>
          ) : null}
        </Fragment>
      ))}
    </tbody>
  );
}
