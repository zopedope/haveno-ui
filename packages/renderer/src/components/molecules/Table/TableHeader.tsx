// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useTableContext } from "./use-table-context";

export function TableHeader() {
  const {
    table,
    props: { showHeader },
  } = useTableContext();

  if (!showHeader) {
    return null;
  }

  return (
    <thead>
      {table.getHeaderGroups().map((headerGroup) => (
        <tr key={headerGroup.id}>
          {headerGroup.headers.map((header) => (
            <th
              key={header.id}
              colSpan={header.colSpan}
              style={{
                width: header.getSize(),
                textAlign: header.column.columnDef?.meta?.textAlign,
              }}
            >
              {header.isPlaceholder ? null : header.renderHeader()}
            </th>
          ))}
        </tr>
      ))}
    </thead>
  );
}
