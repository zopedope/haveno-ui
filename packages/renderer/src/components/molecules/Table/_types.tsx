// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

/* eslint-disable @typescript-eslint/no-explicit-any */
import type { ColumnDef, Row, TableState } from "@tanstack/react-table";
import type { TableProps as MTableProps } from "@mantine/core";

// TODO: Add type or generic
export interface TableProps {
  columns: Array<ColumnDef<any>>;
  table: any;
  data: Array<any>;
  state?: Partial<TableState>;
  defaultColumn?: any;

  showHeader?: boolean;
  showFooter?: boolean;

  rowSubComponent?: ({ row }: { row: Row<any> }) => React.ReactNode;

  tableWrap?: MTableProps;
  variant?: TableVariant;

  onEditableDataChange?: (v: Array<any>) => void;

  pointerRow?: boolean;
  onRowClick?: (column: any) => void;
}

export enum TableVariant {
  Default = "Default",
  Primary = "Primary",
}
