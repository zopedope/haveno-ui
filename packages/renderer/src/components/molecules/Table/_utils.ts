// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

export const updateTableCell = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  data: Array<any>,
  rowIndex: number,
  columnId: string,
  value: unknown
) => {
  return data.map((row, index) => {
    if (index === rowIndex) {
      return {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        ...data[rowIndex]!,
        [columnId]: value,
      };
    }
    return row;
  });
};
