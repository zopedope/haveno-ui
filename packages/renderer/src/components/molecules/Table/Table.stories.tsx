// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { createTable } from "@tanstack/react-table";
import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { Table } from "./Table";

export default {
  title: "atoms/Table",
  component: Table,
} as ComponentMeta<typeof Table>;

const Template: ComponentStory<typeof Table> = () => {
  return (
    <Table
      table={table}
      data={data}
      columns={columns}
      rowSubComponent={() => "asdasd"}
    />
  );
};

export const Default = Template.bind({});

Default.args = {};

interface Person {
  firstName: string;
  lastName: string;
  age: number;
  visits: number;
  status: string;
  progress: number;
}

const table = createTable().setRowType<Person>();

const columns = [
  table.createGroup({
    header: "Name",
    footer: (props) => props.column.id,
    columns: [
      table.createDataColumn("firstName", {
        cell: (info) => info.getValue(),
        footer: (props) => props.column.id,
      }),
      table.createDataColumn((row) => row.lastName, {
        id: "lastName",
        cell: (info) => info.getValue(),
        header: () => <span>Last Name</span>,
        footer: (props) => props.column.id,
      }),
    ],
  }),
  table.createGroup({
    header: "Info",
    footer: (props) => props.column.id,
    columns: [
      table.createDataColumn("age", {
        header: () => "Age",
        footer: (props) => props.column.id,
      }),
      table.createGroup({
        header: "More Info",
        columns: [
          table.createDataColumn("visits", {
            header: () => <span>Visits</span>,
            footer: (props) => props.column.id,
          }),
          table.createDataColumn("status", {
            header: "Status",
            footer: (props) => props.column.id,
          }),
          table.createDataColumn("progress", {
            header: "Profile Progress",
            footer: (props) => props.column.id,
          }),
        ],
      }),
    ],
  }),
];

const data: Array<Person> = [
  {
    firstName: "tanner",
    lastName: "linsley",
    age: 24,
    visits: 100,
    status: "In Relationship",
    progress: 50,
  },
  {
    firstName: "tandy",
    lastName: "miller",
    age: 40,
    visits: 40,
    status: "Single",
    progress: 80,
  },
  {
    firstName: "joe",
    lastName: "dirte",
    age: 45,
    visits: 20,
    status: "Complicated",
    progress: 10,
  },
];
