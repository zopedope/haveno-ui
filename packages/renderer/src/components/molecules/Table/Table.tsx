// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import {
  useTableInstance,
  getCoreRowModel,
  getExpandedRowModel,
} from "@tanstack/react-table";
import { Table as MTable } from "@mantine/core";
import type { TableProps } from "./_types";
import { TableVariant } from "./_types";
import { TableProvider } from "./use-table-context";
import { TableHeader } from "./TableHeader";
import { TableBody } from "./TableBody";
import { useStyles } from "./Table.style";
import { updateTableCell } from "./_utils";

export function Table(props: TableProps) {
  const { classes, cx } = useStyles();
  const {
    table,
    columns,
    data,
    tableWrap,
    variant,
    onEditableDataChange,
    defaultColumn,
    state,
  } = props;

  const tableInstance = useTableInstance(table, {
    data,
    columns,
    state,
    getCoreRowModel: getCoreRowModel(),
    getExpandedRowModel: getExpandedRowModel(),
    meta: {
      updateData: (rowIndex: number, columnId: string, value: unknown) => {
        const newData = updateTableCell(data, rowIndex, columnId, value);
        onEditableDataChange && onEditableDataChange(newData);
      },
    },
    ...(defaultColumn
      ? {
          defaultColumn,
        }
      : {}),
  });

  return (
    <MTable
      {...tableWrap}
      className={cx(tableWrap?.className, {
        [classes.primary]: variant === TableVariant.Primary,
      })}
    >
      <TableProvider value={{ table: tableInstance, props }}>
        <TableHeader />
        <TableBody />
      </TableProvider>
    </MTable>
  );
}

Table.defaultProps = {
  showHeader: true,
  showFooter: true,
};
