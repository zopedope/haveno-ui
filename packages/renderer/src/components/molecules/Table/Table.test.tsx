// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { createTable } from "@tanstack/react-table";
import { Table } from "./Table";

describe("molecules::Table", () => {
  it("renders without exploding.", () => {
    const { asFragment, unmount } = render(
      <Table table={table} data={data} columns={columns} />
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("renders all columns.", () => {
    const { unmount } = render(
      <Table table={table} data={data} columns={columns} />
    );
    expect(screen.queryByText("Name")).toBeInTheDocument();
    expect(screen.queryByText("First Name")).toBeInTheDocument();
    expect(screen.queryByText("Last Name")).toBeInTheDocument();
    unmount();
  });

  it("shouldn't render columns if `showHeader` was true.", () => {
    const { unmount } = render(
      <Table table={table} data={data} columns={columns} showHeader={false} />
    );
    expect(screen.queryByText("Name")).not.toBeInTheDocument();
    expect(screen.queryByText("First Name")).not.toBeInTheDocument();
    expect(screen.queryByText("Last Name")).not.toBeInTheDocument();
    unmount();
  });

  it("renders all columns.", () => {
    const { unmount } = render(
      <Table table={table} data={data} columns={columns} />
    );
    expect(screen.queryByText("Ahmed")).toBeInTheDocument();
    expect(screen.queryByText("Subir")).toBeInTheDocument();
    expect(screen.queryByText("In Relationship")).not.toBeInTheDocument();
    unmount();
  });
});

type Person = {
  firstName: string;
  lastName: string;
  age: number;
  visits: number;
  status: string;
  progress: number;
};
const table = createTable().setRowType<Person>();

const data: Array<Person> = [
  {
    firstName: "Ahmed",
    lastName: "Subir",
    age: 24,
    visits: 100,
    status: "In Relationship",
    progress: 50,
  },
];

const columns = [
  table.createGroup({
    header: "Name",
    footer: (props) => props.column.id,
    columns: [
      table.createDataColumn("firstName", {
        header: "First Name",
        footer: (props) => props.column.id,
      }),
      table.createDataColumn("lastName", {
        id: "lastName",
        header: "Last Name",
        footer: (props) => props.column.id,
      }),
    ],
  }),
];
