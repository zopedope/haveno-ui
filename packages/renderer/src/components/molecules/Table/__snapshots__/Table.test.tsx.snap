// Vitest Snapshot v1

exports[`molecules::Table > renders without exploding. 1`] = `
<DocumentFragment>
  <table
    class="mantine-Table-root mantine-16nz4s2"
  >
    <thead>
      <tr>
        <th
          colspan="2"
          style="width: 300px;"
        >
          Name
        </th>
      </tr>
      <tr>
        <th
          colspan="1"
          style="width: 150px;"
        >
          First Name
        </th>
        <th
          colspan="1"
          style="width: 150px;"
        >
          Last Name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style="width: 150px;"
        >
          Ahmed
        </td>
        <td
          style="width: 150px;"
        >
          Subir
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
